README.md
# Concatenate weather data histories in CSV format collected from OpenWeatherMap for the Belgian communes

The software concatenates the histories Weather data of the Belgian communes collected by using this repository [https://github.com/ALT-F1/OpenWeatherMap](https://github.com/ALT-F1/OpenWeatherMap)

The histories are extracted from the OpenWeatherMap database. See [https://www.openweathermap.org](https://www.openweathermap.org)

The weather data will be linked to other data to answer one question: "Does the usage of the Weather data influence (or not) the model forecasting the spread of the COVID-19?"

It takes 250 seconds to concatenate 742 files in a OpenWeatherMap CSV files on a Intel(R) Core(TM) i7-2640M CPU @ 2.80GHz.

# The context

Given the consequences of the COVID-19 pandemic for public health, many stakeholders in public and private sectors engage in global efforts to treat and understand those exposed to the virus and to contain the outbreak.

Universities and private companies contribute to the analysis of data useful to forecast the spread of the "Severe Acute Respiratory Syndrome CoronaVirus 2" (SARS-CoV-2).

Those persons include researchers and consultants who share their analysis to the public administration and members of governments who will shape the rules and regulations describing how the citizens should behave during the pandemic.

# Why collecting the weather data?

In Belgium, the  Group of Experts for an Exit Strategy (GEES) drafts proposals to envisage a gradual deconfinement. The GEES supports the [Belgian Federal Government](https://www.belgium.be/en/about_belgium/government/federal_authorities/federal_government).

See also [https://www.info-coronavirus.be/en](https://www.info-coronavirus.be/en)

The GEES uses data from researchers, amongst others. Those researchers build models of the spread of the virus by using diverse data: mobility, health, beds in hospitals, fatalities, current regulation ...

One question remains: **does the usage of the Weather data influence (or not) the spread of the virus?**

One of those groups of researchers includes the [Machine Learning Group](https://mlg.ulb.ac.be) (MLG) from the [Université Libre de Bruxelles](https://www.ulb.be). The MLG supports the initiative by providing models made of Artificial Intelligence/Machine Learning algorithms. The models built by the MLG influence the Exit Strategy.

# The tools used to build the software

* Pandas https://pandas.pydata.org
* Python https://www.python.org 
* Numpy https://numpy.org
* Kaggle https://www.kaggle.com
* Visual Studio Code https://code.visualstudio.com

# The sponsors of the project

The **EOSC Secretariat** supports the governance of the **E**uropean **O**pen **S**cience **C**loud (EOSC). See [https://www.eoscsecretariat.eu/](https://www.eoscsecretariat.eu)

The EOSC supports projects aiming to make data **F**indable, **A**ccessible, **I**nteroperable, and **R**eproducible (FAIR) for scientists; these combinations would lead to (unforeseen) reuse and faster development of science. 

EOSC Secretariat granted funds to [http://www.alt-f1.be](http://www.alt-f1.be) to create a "Crossroads Data Bank for COVID-19".

Latest news from the European Commission: [https://ec.europa.eu/research/openscience/index.cfm?pg=open-science-cloud](https://ec.europa.eu/research/openscience/index.cfm?pg=open-science-cloud) 

**OpenWeatherMap** provides, free-of-charge, access to weather data to its historical weather data [https://openweathermap.org/api](https://openweathermap.org/api)

OpenWeather Ltd is a British-based tech company that provides weather and satellite data worldwide. OpenWeather collects and processes raw data from a variety of sources, and gives its customers access to the archive. See [https://openweathermap.org/](https://openweathermap.org/)

See [Openweather LTD Helps The Fight To Overcome COVID-19](https://bit.ly/2ZohgFF)

# The supporters of the project

**The Machine Learning Group**, founded in 2004 by [Gianluca Bontempi](https://mlg.ulb.ac.be/wordpress/members-2/gianluca-bontempi),  is a research unit of the Computer Science Department of the ULB (Université Libre de Bruxelles, Brussels, Belgium), Faculty of Sciences, currently co-headed by Prof. Gianluca Bontempi and Prof. [Tom Lenaerts](https://mlg.ulb.ac.be/wordpress/members-2/tom-lenaerts).

MLG targets machine learning and behavioral intelligence research focusing on time series analysis, big data mining, causal inference, network inference, decision-making models, and behavioral analysis with applications in data science, medicine, molecular biology, cybersecurity and social dynamics related to cooperation, emotions, and others.

See [https://mlg.ulb.ac.be/](https://mlg.ulb.ac.be/)

# Legal matters

The current software digests data that do not contain personal data.

The output of this project is linked to data that do not contain Personally Identifiable Information (PII). See [https://en.wikipedia.org/wiki/Personal_data](https://en.wikipedia.org/wiki/Personal_data)

# the data models

## OpenWeatherMap Historical data

OpenWeatherMap provides an API returning the hourly Weather. A call to the API returns a maximum of 7 consecutive days of data.
[https://openweathermap.org/history](https://openweathermap.org/history) 

Here is an example of a call: `http://history.openweathermap.org/data/2.5/history/city?id={id}&type=hour&start={start}&end={end}&appid={YOUR_API_KEY}`

### the response from the history of OpenWeatherMap

The documentation of the **response from the history** of OpenWeatherMap is here: [https://openweathermap.org/weather-data](https://openweathermap.org/weather-data)

``` json
{
    "dt": 1587312000,
    "main": {
        "temp": 287.44,
        "feels_like": 284.3,
        "pressure": 1015,
        "humidity": 76,
        "temp_min": 287.04,
        "temp_max": 288.15
    },
    "wind": {
        "speed": 4.6,
        "deg": 40
    },
    "clouds": {
        "all": 77
    },
    "weather": [
        {
            "id": 500,
            "main": "Rain",
            "description": "light rain",
            "icon": "10d"
        }
    ],
    "rain": {
        "1h": 0.3
    }
},
```
## OpenWeatherMap historical data in CSV format

The project [https://github.com/ALT-F1/OpenWeatherMap](https://github.com/ALT-F1/OpenWeatherMap) generates either CSV or JSON files containing the Weather of one Belgian city per day hour by hour.

The current project concatenates the historical data into one single file.

Historical data collected from March 01, 2020, to May 22, 2020, weights 231,9 MB in a CSV format and 251,9 MB in a pickle format.

CSV and Pickle files are available in `src/kaggle/output/kaggle/working`.

Here is an extract of the CSV format:

``` csv
message,cod,city_id,calctime,cnt,dt,main.temp,main.feels_like,main.pressure,main.humidity,main.temp_min,main.temp_max,wind.speed,wind.deg,clouds.all,weather.id,weather.main,weather.description,weather.icon

Count: 168,200,2787356,0.005035309,168,1583017200,278.11,271.54,995,86,277.04,279.15,7.2,210,40,802,Clouds,scattered clouds,03n

Count: 168,200,2787356,0.005035309,168,1583020800,278.31,270.13,995,80,277.04,279.82,9.3,210,20,801,Clouds,few clouds,02n

Count: 168,200,2787356,0.005035309,168,1583024400,278.31,270.4,995,75,277.04,279.82,8.7,210,20,801,Clouds,few clouds,02n
```

The concatenation of the historical data augmented by some transformations generate this format:

``` csv
,message,cod,city_id,calctime,cnt,dt,main.temp,main.feels_like,main.pressure,main.humidity,main.temp_min,main.temp_max,wind.speed,wind.deg,clouds.all,weather.id,weather.main,weather.description,weather.icon,Province,Postal code,date

0,Count: 168,200,2795949,0.037120004,168,1583017200,278.15,271.58,995,86,277.04,279.26,7.2,210,40,802,Clouds,scattered clouds,03n,LIEGE,4217,2020-02-29

1,Count: 168,200,2795949,0.037120004,168,1583020800,278.22,270.02,995,80,277.04,279.26,9.3,210,87,500,Rain,light rain,10n,LIEGE,4217,2020-03-01

2,Count: 168,200,2795949,0.037120004,168,1583024400,278.25,270.33,995,75,277.04,279.26,8.7,210,66,500,Rain,light rain,10n,LIEGE,4217,2020-03-01

```


# How to contribute to software development

* run `npm install` to install the standard-conventions package. A utility for versioning using semver [https://semver.org/](https://semver.org/) and CHANGELOG generation powered by Conventional Commits [https://conventionalcommits.org](https://conventionalcommits.org).

* commit your changes by using the scripts. read the specifications [https://www.conventionalcommits.org/en/v1.0.0/#specification](https://www.conventionalcommits.org/en/v1.0.0/#specification)
    * run `git commit -m "{build}{chore}{ci}{docs}{feat}{fix}{perf}{refactor}{revert}{style}{test}: xxx"`

* set the version of the code and amend the CHANGELOG with
    * `npm run patch` (from 0.0.0 to 0.0.1)
    * `npm run minor` (from 0.0.0 to 0.1.0)
    * `npm run major` (from 0.0.0 to 1.0.0)

* push the code including the tags
    * `npm run push`


# Acronyms

* ETL: Extract-Transform-Load https://en.wikipedia.org/wiki/Extract,_transform,_load

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.0.0](https://github.com/ALT-F1/OpenWeatherMap/compare/v1.1.0...v2.0.0) (2020-06-08)


### Features

* build dataframes grouping the weather data by province, date and including the quartiles ([b16deac](https://github.com/ALT-F1/OpenWeatherMap/commit/b16deac5c241d8f1f37e50c1978152fb3d13d574))

## 1.1.0 (2020-06-08)


### Features

* add alt-be.helper to check if Kaggle is running or not ([6905478](https://github.com/ALT-F1/OpenWeatherMap/commit/690547878b67877c8cf0dade202bc8b3583e8418))
* add OpenWeatherMap.org data in CSV format ([1b7b140](https://github.com/ALT-F1/OpenWeatherMap/commit/1b7b140ea242f77668fa879a9bdc6c99cbbaadc1))
* add zipcodes from Belgium in FR and NL to find the Provinces related to Postal Codes ([302427c](https://github.com/ALT-F1/OpenWeatherMap/commit/302427c4961702153c968157fd4e599834d13198))
* aggregate CSV files containing Weather data per day into a CSV and Pickle file ([89cc85f](https://github.com/ALT-F1/OpenWeatherMap/commit/89cc85fb3ad8be1dae81416d49e5699ef83f5f74))
* prepare the Belgian postal codes database from BPost before ETL ([d0f61c1](https://github.com/ALT-F1/OpenWeatherMap/commit/d0f61c1629f5a9f2679ffcfcd28cf0de1e4dbbea))


### Documentations

* add a closed LICENSE ([50c4c64](https://github.com/ALT-F1/OpenWeatherMap/commit/50c4c64367fb07d172e845ff2405cc929bcd4893))
* add README ([3dc5b3b](https://github.com/ALT-F1/OpenWeatherMap/commit/3dc5b3b4ba48f0b2000144789fabc190859708f2))
* remove some text ([632b916](https://github.com/ALT-F1/OpenWeatherMap/commit/632b916872ece536eb636c5e8115a04379641931))


### Chores

* add .gitattributes including csv files ([61e9d73](https://github.com/ALT-F1/OpenWeatherMap/commit/61e9d73a7e11f955e0d19cca6c6492433dc7bf4d))
* add .gitignore for node, python, jupyter notebooks, visual studio code ([d203d3d](https://github.com/ALT-F1/OpenWeatherMap/commit/d203d3d8a6a9f8a522c346a0f3328b9528a521b2))
* add vscode configurations ([e1443ce](https://github.com/ALT-F1/OpenWeatherMap/commit/e1443ce5d6383de7ccd692a685aa0a76296d122e))
* **release:** 1.0.0 ([dcb85e1](https://github.com/ALT-F1/OpenWeatherMap/commit/dcb85e14da219a26ea3174fccc97a685be3e2fe2))
* ignore files greater than 100 MB due to gihutb constraint : remove df_provinces_collection.pkl and df_provinces_collection.csv ([87f27d9](https://github.com/ALT-F1/OpenWeatherMap/commit/87f27d93063a4e0190ed2c9a738eb7520809ae5e))
* set the types recognized during the generation of the CHANGELOG ([39f01a9](https://github.com/ALT-F1/OpenWeatherMap/commit/39f01a90543c1a49fa0dd9f13e92b8c6b86284aa))
* **release:** 1.0.1 ([c081503](https://github.com/ALT-F1/OpenWeatherMap/commit/c081503754a60a24a9f8cf8d48eb398fc712623c))


### Tests

* add concatenated CSV and Pickle files containing all Weather data from Belgian cities ([735ee0e](https://github.com/ALT-F1/OpenWeatherMap/commit/735ee0ea968b9035680fb779e5bd569b99bcec78))
* add zipped version of df_provinces_collection.pkl and df_provinces_collection.csv ([b6c97ac](https://github.com/ALT-F1/OpenWeatherMap/commit/b6c97ac26b25441a9994d64f048299b8884f852f))


### Builds

* add package.json ([a3c5eeb](https://github.com/ALT-F1/OpenWeatherMap/commit/a3c5eeb64c17b8b44b4ff95d3d3e71f85ba9e673))
* add requirements.txt ([f5cfdaa](https://github.com/ALT-F1/OpenWeatherMap/commit/f5cfdaafe72c81142c6548b3ca0a50c40e733ac7))


### Styles

* add the github social preview ([5d23a71](https://github.com/ALT-F1/OpenWeatherMap/commit/5d23a717803d95f2e4d172b5de7ee4e87ecf9061))
* add the github social preview ([d516523](https://github.com/ALT-F1/OpenWeatherMap/commit/d516523843154e845cd16190d30690ffff3847fd))

### [1.0.1](https://github.com/ALT-F1/OpenWeatherMap/compare/v1.0.0...v1.0.1) (2020-06-05)


### Tests

* add zipped version of df_provinces_collection.pkl and df_provinces_collection.csv ([13ed0cb](https://github.com/ALT-F1/OpenWeatherMap/commit/13ed0cb44aec8b4c81a04bafedebaa2c6d9c05d8))


### Chores

* ignore files greater than 100 MB due to gihutb constraint : remove df_provinces_collection.pkl and df_provinces_collection.csv ([67aef5b](https://github.com/ALT-F1/OpenWeatherMap/commit/67aef5ba1e0ded5e52acc644d86c70cda652ad18))

## 1.0.0 (2020-06-05)


### Features

* add alt-be.helper to check if Kaggle is running or not ([6905478](https://github.com/ALT-F1/OpenWeatherMap/commit/690547878b67877c8cf0dade202bc8b3583e8418))
* add OpenWeatherMap.org data in CSV format ([1b7b140](https://github.com/ALT-F1/OpenWeatherMap/commit/1b7b140ea242f77668fa879a9bdc6c99cbbaadc1))
* add zipcodes from Belgium in FR and NL to find the Provinces related to Postal Codes ([302427c](https://github.com/ALT-F1/OpenWeatherMap/commit/302427c4961702153c968157fd4e599834d13198))
* aggregate CSV files containing Weather data per day into a CSV and Pickle file ([89cc85f](https://github.com/ALT-F1/OpenWeatherMap/commit/89cc85fb3ad8be1dae81416d49e5699ef83f5f74))
* prepare the Belgian postal codes database from BPost before ETL ([d0f61c1](https://github.com/ALT-F1/OpenWeatherMap/commit/d0f61c1629f5a9f2679ffcfcd28cf0de1e4dbbea))



src/eosc-gees-openweathermap_by_province_in_be.py
# %% [code]
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import pandas as pd
from altf1be_helpers import AltF1BeHelpers
from bpost_be_postal_code_helpers import BPost_postal_codes
import time
import datetime
import numpy as np  # linear algebra
import pandas as pd  # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 5GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All"
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

# %% [markdown]
# # FEAT: requirements
#
# Pour gerer la variabilité on pourrait stocker par jour plusieurs quantiles de la variable: par exemple si en province de Antwerp on prend 10 villes, on stocke les quantiles à 25-50-75% . L’important est que j’aie une valeur par jour et par province
#
# See https://covid19analyt-oho5182.slack.com/archives/C013F1FUDME/p1590487917005000

# %% [code]
# import libraries

# %% [code]
# Set the directories where the extract are

history_city_dir = "/kaggle/input/openweathermap-historical"

# %% [code]
# instanciate required classes

bpost_postal_codes = BPost_postal_codes()
df_postal_codes_in_be = bpost_postal_codes.df_postal_codes_in_be

# %% [markdown]
# # TRANSFORM:Translate the provinces from NL to FR

# %% [code]


def translate_provinces_in_french(df_postal_codes_in_be):
    df_postal_codes_in_be.loc[df_postal_codes_in_be['Province'] == 'ANTWERPEN', [
        'Province']] = 'ANVERS'
    df_postal_codes_in_be.loc[df_postal_codes_in_be['Province'] == 'BRUSSEL', [
        'Province']] = 'BRUXELLES'
    df_postal_codes_in_be.loc[df_postal_codes_in_be['Province'] == 'HENEGOUWEN', [
        'Province']] = 'HAINAUT'
    df_postal_codes_in_be.loc[df_postal_codes_in_be['Province'] == 'LIMBURG', [
        'Province']] = 'LIMBOURG'
    df_postal_codes_in_be.loc[df_postal_codes_in_be['Province'] == 'LUIK', [
        'Province']] = 'LIEGE'
    df_postal_codes_in_be.loc[df_postal_codes_in_be['Province'] == 'LUXEMBURG', [
        'Province']] = 'LUXEMBOURG'
    df_postal_codes_in_be.loc[df_postal_codes_in_be['Province'] == 'NAMEN', [
        'Province']] = 'NAMUR'
    df_postal_codes_in_be.loc[df_postal_codes_in_be['Province']
                              == 'OOST-VLAANDEREN', ['Province']] = 'FLANDRE-ORIENTALE'
    df_postal_codes_in_be.loc[df_postal_codes_in_be['Province']
                              == 'WEST-VLAANDEREN', ['Province']] = 'FLANDRE-OCCIDENTALE'
    df_postal_codes_in_be.loc[df_postal_codes_in_be['Province']
                              == 'VLAAMS-BRABANT', ['Province']] = 'BRABANT FLAMAND'
    df_postal_codes_in_be.loc[df_postal_codes_in_be['Province']
                              == 'WAALS-BRABANT', ['Province']] = 'BRABANT WALLON'
    return df_postal_codes_in_be

# %% [code]
# group by province


df_postal_codes_in_be = translate_provinces_in_french(df_postal_codes_in_be)
df_grouped_by_province = df_postal_codes_in_be.groupby(['Province'])[
    'Code postal'].apply(set)
df_grouped_by_province

# %% [markdown]
# # INITIALIZATIONS: filenames pattern

# %% [code]
openweathermap_org_weather_directory = "kaggle/input/openweathermap-org/"

# %% [code]
df_postal_codes_in_be.columns

# %% [code]
# concatenate the cities in each province


def concatenate_cities_in_each_province(postal_codes, year=2020, months=[3]):
    df = pd.DataFrame()

    first_day_of_the_month, last_day_of_the_month = AltF1BeHelpers.get_first_and_last_day_of_the_month(
        year=year, month=months[len(months)-1])

    for city_postal_code in postal_codes:
        print(f"city_postal_code: {city_postal_code}")
        city_name = df_postal_codes_in_be.loc[df_postal_codes_in_be['Code postal']
                                              == city_postal_code]['Commune principale normalized']
        print(f"Commune city_name: {city_name}")
        city_name = df_postal_codes_in_be.loc[df_postal_codes_in_be['Code postal']
                                              == city_postal_code]['Localité normalized']
        print(f"Localité city_name: {city_name}")

        filename = (f'{city_postal_code.zfill(4)}_{city_name}_from_{first_day_of_the_month.strftime("%Y-%m-%d_%Hh%M")}_to_{last_day_of_the_month.strftime("%Y-%m-%d_%Hh%M")}-{first_day_of_the_month.strftime("%s")}_to_{last_day_of_the_month.strftime("%s")}')
        print(f"filename: {filename}")
        csv_path = os.path.join(openweathermap_org_weather_directory,
                                f'{city_postal_code.zfill(4)}_{city_name}_from_{first_day_of_the_month.strftime("%Y-%m-%d_%Hh%M")}_to_{last_day_of_the_month.strftime("%Y-%m-%d_%Hh%M")}-{first_day_of_the_month.strftime("%s")}_to_{last_day_of_the_month.strftime("%s")}')
        print(f"{csv_path}")
        df_new_cities_per_province = pd.read_csv(f"{csv_path}.csv", sep=',')

        df = pd.concat([df, df_postal_codes_in_be])
    df
    return df

# %% [code]
# compute the quantiles

# %% [markdown]
# # LOAD : for each Postal code we look for the quantile

# %% [code]
# LOAD: for each file in input/openweathermap_org


# %% [code]
df_grouped_by_province.index

# %% [code]
PROVINCE_NOT_FOUND = -1  # const


def get_province_from(postal_code):
    """
    if the Belgian province for a postal code
    """
    index = 0
    # print(type(df_grouped_by_province))
    for cities_in_province in df_grouped_by_province:
        is_postal_code_in_list = True if postal_code in cities_in_province else False
        if is_postal_code_in_list:
            # print(df_grouped_by_province)
            # print(type(cities_in_province))
            # print(f"{cities_in_province}")
            # print(f"{is_postal_code_in_list}")
            # print(f"{df_grouped_by_province.index}")
            return df_grouped_by_province.index[index]
        index = index + 1

    return PROVINCE_NOT_FOUND


# %% [code]
# keep: test from which province comes this postal code
print(get_province_from(1932))

# %% [code]


def get_normalized_province(province):
    return province.replace('-', '')

# %% [code]
# keep the code


def append_df_per_province(df, filename):
    postal_code = filename[:4]
    province = get_province_from(postal_code=int(postal_code))
    df['Province'] = province
    df['Postal code'] = postal_code
    df['date'] = df['dt'].apply(
        lambda x: time.strftime('%Y-%m-%d', time.localtime(x)))
    print(f"{postal_code}-{province}")
    # print(df)
    # df.columns
    #current_province = df_postal_codes_in_be.loc[df_postal_codes_in_be['Code postal'] == df.loc['Province']]
    return df


# %% [code]
# keep the code: create one DataFrame containing all CSV containing March 01 to May 22, 2020
t0 = time.clock()
index = 0
df_provinces_collection = pd.DataFrame()  # one dataframe per province

dir_data = os.path.join(os.path.abspath(
    os.getcwd()), "src", openweathermap_org_weather_directory
)
for dirname, _, filenames in os.walk(dir_data):
    for filename in filenames:
        print(f"loading: {index}/{len(filenames)}")
        csv_path = os.path.join(dirname, filename)
        print(filename)
        df_provinces_collection = pd.concat(
            [df_provinces_collection, append_df_per_province(
                pd.read_csv(csv_path, sep=','), filename)]
        )
        index = index+1
t1 = time.clock() - t0
print("Time elapsed: ", t1 - t0)  # CPU seconds elapsed (floating point)
print("finished\n")
df_provinces_collection

# %% [code]
# store current dataframe containing the weather of all cities from March 01 to May 22, 2020
df_provinces_collection.to_csv("df_provinces_collection.csv")
# too long df_provinces_collection.to_excel("df_provinces_collection.xlsx")
df_provinces_collection.to_pickle("df_provinces_collection.pkl")
# with pd.HDFStore('df_provinces_collection_01.h5', mode='w') as f:
#    f.append(df_provinces_collection, key='df', format='fixed', data_columns=df_provinces_collection.columns)

#df_provinces_collection.to_hdf("df_provinces_collection.hdf5", format='table', key='/provinces')

# %% [code]

df_quantiles = df_provinces_collection.copy(deep=True)

# %% [code]

df_quantiles.columns

# %% [code]
# keep: add a column with a date based on dt
df_quantiles['date'] = df_quantiles['dt'].apply(
    lambda x: time.strftime('%Y-%m-%d', time.localtime(x)))

# %% [code]
df_quantiles

# %% [code]
total = 0
for province in df_quantiles['Province'].unique():
    total = total + df_quantiles[df_quantiles['Province'] == province].shape[0]
    print(
        f"{province}-{df_quantiles[df_quantiles['Province'] == province].shape}")
print(f"{total}")

# %% [code]
current_province = df_quantiles.loc[df_quantiles['Province'] == "BRUXELLES"]
current_province

# %% [code]
current_province.columns

# %% [code]
current_date_df = current_province.loc[current_province['date']
                                       == '2020-03-01']
current_date_df.head(3)

# %% [code]
current_province.shape

# %% [code]
new_df = pd.DataFrame(columns=['message', 'cod', 'city_id', 'dt', 'main.temp25',
                               'main.temp50', 'main.temp75',
                               'main.feels_like25', 'main.feels_like50', 'main.feels_like75',
                               'main.pressure25', 'main.pressure50', 'main.pressure75',
                               'main.humidity25', 'main.humidity50', 'main.humidity75',
                               'main.temp_min25', 'main.temp_min50', 'main.temp_min75',
                               'main.temp_max25', 'main.temp_max50', 'main.temp_max75',
                               'wind.speed25', 'wind.speed50', 'wind.speed75',
                               'wind.deg25', 'wind.deg50', 'wind.deg75',
                               'Province',
                               'date'])
# 'clouds.all', 'weather.id',
#'weather.main', 'weather.description', 'weather.icon',

# %% [code]
current_date_df['main.temp'].quantile(.25)

# %% [code]
new_df.loc[0, 'main.temp25'] = current_date_df['main.temp'].quantile(.25)
new_df.loc[0, 'main.temp50'] = current_date_df['main.temp'].quantile(.50)

# %% [code]
new_df

# %% [code]


def add_quantiles(new_df):
    # per province
    columns = ['message', 'cod', 'dt',
               'Province',
               'date',
               'main.temp25', 'main.temp50', 'main.temp75',
               'main.feels_like25', 'main.feels_like50', 'main.feels_like75',
               'main.pressure25', 'main.pressure50', 'main.pressure75',
               'main.humidity25', 'main.humidity50', 'main.humidity75',
               'main.temp_min25', 'main.temp_min50', 'main.temp_min75',
               'main.temp_max25', 'main.temp_max50', 'main.temp_max75',
               'wind.speed25', 'wind.speed50', 'wind.speed75',
               'wind.deg25', 'wind.deg50', 'wind.deg75',
               ]
    new_df = pd.DataFrame(columns=columns)
    lines = 0

    df_quantiles_sorted = df_quantiles.sort_values(['Province', 'dt'])

    for province in df_quantiles_sorted['Province'].unique():
        # per day
        # print(f"************************************************************************")
        # print(f"************************************************************************")
        # print(f"************************************************************************")
        print(f"************************************************************************")

        print(f"PROVINCE: {province}")
        for date in df_quantiles_sorted['date'].unique():
            print(f"LINES: {lines} - {new_df.shape}")
            lines = lines+1
            print(
                f"************************************************************************")
            # print(f"************************************************************************")
            # print(f"************************************************************************")
            print(f"DATE : {date}")
            #new_row = pd.DataFrame(columns=columns)
            # compute quantile
            current_date_df = df_quantiles_sorted.loc[df_quantiles_sorted['date'] == date]

            new_row = pd.DataFrame([[
                current_date_df.iloc[0]['message'], current_date_df.iloc[0]['cod'], current_date_df.iloc[0]['dt'], province, date, current_date_df['main.temp'].quantile(.25), current_date_df['main.temp'].quantile(.50), current_date_df['main.temp'].quantile(.75), current_date_df['main.feels_like'].quantile(.25), current_date_df['main.feels_like'].quantile(.50), current_date_df['main.feels_like'].quantile(.75), current_date_df['main.pressure'].quantile(.25), current_date_df['main.pressure'].quantile(.50), current_date_df['main.pressure'].quantile(.75), current_date_df['main.humidity'].quantile(.25), current_date_df['main.humidity'].quantile(
                    .50), current_date_df['main.humidity'].quantile(.75), current_date_df['main.temp_min'].quantile(.25), current_date_df['main.temp_min'].quantile(.50), current_date_df['main.temp_min'].quantile(.75), current_date_df['main.temp_max'].quantile(.25), current_date_df['main.temp_max'].quantile(.50), current_date_df['main.temp_max'].quantile(.75), current_date_df['wind.speed'].quantile(.25), current_date_df['wind.speed'].quantile(.50), current_date_df['wind.speed'].quantile(.75), current_date_df['wind.deg'].quantile(.25), current_date_df['wind.deg'].quantile(.50), current_date_df['wind.deg'].quantile(.75)
            ]],
                columns=columns)

            #new_row = pd.DataFrame([date], columns=['message'])
            print(f"new_row.shape : {new_row.shape}")
            print(f"new_row : {new_row}")
            #print(f"new_row.columns : {new_row.columns}")
            new_df = new_df.append(new_row, ignore_index=True)
    return new_df

# %% [markdown]
# # load the df_provinces_collection and work


# %% [code]
df_with_quantiles = pd.DataFrame()
df_with_quantiles = add_quantiles(df_with_quantiles)

# %% [code]
df_with_quantiles.shape

# %% [code]
df_with_quantiles

# %% [code]
df_with_quantiles_cleaned = df_with_quantiles[df_with_quantiles['Province'] != -1]

# %% [code]
df_with_quantiles_cleaned.shape

# %% [code]
df_with_quantiles.to_csv(
    "openweather_map-belgium-per_province_per_day_with_quantiles.csv")
df_with_quantiles.to_excel(
    "openweather_map-belgium-per_province_per_day_with_quantiles.xlsx")

# %% [code]
df_with_quantiles_cleaned.to_csv(
    "openweather_map-belgium-per_province_per_day_with_quantiles_cleaned.csv")
df_with_quantiles_cleaned.to_excel(
    "openweather_map-belgium-per_province_per_day_with_quantiles_cleaned.xlsx")

# %% [markdown]
# # FINISH - THIS IS THE END

# %% [code]
df_quantiles_sorted = df_quantiles.sort_values(['Province', 'dt'])

# %% [code]
df_quantiles_sorted.shape

# %% [code]
province = "ANVERS"
df_for_one_day_one_province = df_quantiles_sorted.loc[df_quantiles_sorted['Province'] == province]

# %% [code]
df_for_one_day_one_province.shape

# %% [code]
date = "2020-03-01"
df_current_date = df_for_one_day_one_province.loc[df_for_one_day_one_province['date'] == date]

# %% [code]
df_current_date.shape

# %% [code]
current_date_df['main.temp'].quantile(.25)

# %% [code]
columns = ['message', 'cod', 'dt',
           'Province',
           'date',
           'main.temp25', 'main.temp50', 'main.temp75',
           'main.feels_like25', 'main.feels_like50', 'main.feels_like75',
           'main.pressure25', 'main.pressure50', 'main.pressure75',
           'main.humidity25', 'main.humidity50', 'main.humidity75',
           'main.temp_min25', 'main.temp_min50', 'main.temp_min75',
           'main.temp_max25', 'main.temp_max50', 'main.temp_max75',
           'wind.speed25', 'wind.speed50', 'wind.speed75',
           'wind.deg25', 'wind.deg50', 'wind.deg75',
           ]
new_row = pd.DataFrame([[
    current_date_df.iloc[0]['message'], current_date_df.iloc[0]['cod'], current_date_df.iloc[0]['dt'], province, date, current_date_df['main.temp'].quantile(.25), current_date_df['main.temp'].quantile(.50), current_date_df['main.temp'].quantile(.75), current_date_df['main.feels_like'].quantile(.25), current_date_df['main.feels_like'].quantile(.50), current_date_df['main.feels_like'].quantile(.75), current_date_df['main.pressure'].quantile(.25), current_date_df['main.pressure'].quantile(.50), current_date_df['main.pressure'].quantile(.75), current_date_df['main.humidity'].quantile(.25), current_date_df['main.humidity'].quantile(
        .50), current_date_df['main.humidity'].quantile(.75), current_date_df['main.temp_min'].quantile(.25), current_date_df['main.temp_min'].quantile(.50), current_date_df['main.temp_min'].quantile(.75), current_date_df['main.temp_max'].quantile(.25), current_date_df['main.temp_max'].quantile(.50), current_date_df['main.temp_max'].quantile(.75), current_date_df['wind.speed'].quantile(.25), current_date_df['wind.speed'].quantile(.50), current_date_df['wind.speed'].quantile(.75), current_date_df['wind.deg'].quantile(.25), current_date_df['wind.deg'].quantile(.50), current_date_df['wind.deg'].quantile(.75)
]],
    columns=columns)

# %% [code]
new_row.columns

# %% [code]
new_row

# %% [code]
new_row.shape

# %% [code]
new_df = pd.DataFrame(columns=columns)
new_df = new_df.append(new_row, verify_integrity=True,
                       ignore_index=True, sort=False)
new_df.shape

# %% [code]
new_df

# %% [code]
new_df.columns

# %% [code]

df1 = pd.DataFrame({'Name': ['Pankaj', 'Lisa'], 'ID': [1, 2]})
df2 = pd.DataFrame({'Name': ['David'], 'ID': [3]})

print(df1)
print(df2)

df3 = df1.append(df2)
print('\nResult DataFrame:\n', df3)

# %% [code]
df_final = pd.DataFrame(columns=['main.temp25'])

# %% [code]
df_final['main.temp25'] = current_date_df['main.temp'].quantile(.25)

# %% [code]
df_final.shape

# %% [code]
current_date_df['main.temp'].quantile(.50)

# %% [code]


# %% [code]
df_final['main.temp50'] = current_date_df['main.temp'].quantile(.50)
current_date_df['main.temp75'] = current_date_df['main.temp'].quantile(.75)

current_date_df['main.feels_like25'] = current_date_df['main.feels_like'].quantile(
    .25)
current_date_df['main.feels_like50'] = current_date_df['main.feels_like'].quantile(
    .50)
current_date_df['main.feels_like75'] = current_date_df['main.feels_like'].quantile(
    .75)

current_date_df['main.pressure25'] = current_date_df['main.pressure'].quantile(
    .25)
current_date_df['main.pressure50'] = current_date_df['main.pressure'].quantile(
    .50)
current_date_df['main.pressure75'] = current_date_df['main.pressure'].quantile(
    .75)

current_date_df['main.humidity25'] = current_date_df['main.humidity'].quantile(
    .25)
current_date_df['main.humidity50'] = current_date_df['main.humidity'].quantile(
    .50)
current_date_df['main.humidity75'] = current_date_df['main.humidity'].quantile(
    .75)

current_date_df['main.temp_min25'] = current_date_df['main.temp_min'].quantile(
    .25)
current_date_df['main.temp_min50'] = current_date_df['main.temp_min'].quantile(
    .50)
current_date_df['main.temp_min75'] = current_date_df['main.temp_min'].quantile(
    .75)

current_date_df['main.temp_max25'] = current_date_df['main.temp_max'].quantile(
    .25)
current_date_df['main.temp_max50'] = current_date_df['main.temp_max'].quantile(
    .50)
current_date_df['main.temp_max75'] = current_date_df['main.temp_max'].quantile(
    .75)

current_date_df['wind.speed25'] = current_date_df['wind.speed'].quantile(.25)
current_date_df['wind.speed50'] = current_date_df['wind.speed'].quantile(.50)
current_date_df['wind.speed75'] = current_date_df['wind.speed'].quantile(.75)

current_date_df['wind.deg25'] = current_date_df['wind.deg'].quantile(.25)
current_date_df['wind.deg50'] = current_date_df['wind.deg'].quantile(.50)
current_date_df['wind.deg75'] = current_date_df['wind.deg'].quantile(.75)

# %% [code]
current_date_df.columns

# %% [code]
